frontend/src/dashboard.js

```diff
@@ -46,623 +44,528 @@ import {
-  }
+  const navigate = useNavigate();
+
+  // dashboard states
+  const [dashboards, setDashboards] = useState([]);
+  const [filteredDBs, setFilteredDBs] = useState([]);
+  const [selectedDB, setSelectedDB] = useState();
+  const [isDBSelectorOpen, setIsDBSelectorOpen] = useState(false);
+  const [isNewDBOpen, setIsNewDBOpen] = useState(false);
+  const [isDeleteDBOpen, setIsDeleteDBOpen] = useState(false);
+
+  // widget states
+  const [widgets, setWidgets] = useState([]);
+  const [isNewWidgetOpen, setIsNewWidgetOpen] = useState(false);
+  const [isEditModalOpen, setIsEditModalOpen] = useState(false);
+  const [editWidgetData, setEditWidgetData] = useState({});
+  const [isDeleteWidgetOpen, setIsDeleteWidgetOpen] = useState(false);
+  const [currentWidget, setCurrentWidget] = useState();
+
+  // typeahead input value states
+  const [selectDBInputValue, setSelectDBInputValue] = useState('');
+  const [filterDBValue, setFilterDBValue] = useState('');
+
+  useEffect(() => {
+    // fetch widgets with the selected dashboard changes
+
+    // TODO is sync necessary with functional component...?
+    //sync_context();
+    console.log('dashboard useEffect selectedDB');
+    //getDefaultDashboard();  // TODO: only changes on project selection, not needed on setup?
+
+    getWidgets();
+
+  }, [selectedDB]);
+
+  useEffect(() => {
+    // update dashboards when the filter input changes
+    getDashboards();
+    console.log('dashboard useEffect filterDBValue');
+
+  }, [filterDBValue]);
+
+  const { defaultDashboard } = context;
 
-  sync_context = () => {
-    // Active dashboard
-    const { activeDashboard, setActiveDashboard } = this.context;
-    const { selectedDashboard } = this.state;
-    const paramDash = this.props.params?.dashboard_id;
-    if (!paramDash) {
-      // No dashboard in the URL, clear context
-      setActiveDashboard();
+  useEffect(() => {
+    // if there is a defaultDashboard and it's in the current list of dashboards, apply it
+    // implicit attachment to project selection within the context, as that will change defaultDashboard
+    // TODO: account for 'last selected dashboard' preference
+    console.log('dashboard useEffect defaultDashboard');
+
+    if (defaultDashboard ) {
+      const default_db_item = dashboards.filter(dash => dash.id == defaultDashboard).pop();
+      if (default_db_item) {
+        setSelectedDB(default_db_item);
+        setSelectDBInputValue(default_db_item.title);
+      }
     }
-    let updatedDash = undefined;
-    // API call to update context
-    if ( paramDash != null && activeDashboard?.id !== paramDash) {
-      HttpClient.get([Settings.serverUrl, 'dashboard', paramDash])
+  }, [defaultDashboard])
+
+  const { dashboard_id } = params;
+  useEffect(() => {
+    // sync the URL to the dashboard selection
+    if (selectedDB?.id !== dashboard_id) {
+      console.log('dashboard useEffect dashboard_id');
+      HttpClient.get([Settings.serverUrl, 'dashboard', dashboard_id])
```

Comment:
Not so much something to fix, but just a pattern you can use. 
Self evoked functions allow for async usage ie you could do:

(async()=>{
try {
const result = await HttpClient.get([Settings.serverUrl, 'dashboard', dashboard_id])
// do things
} catch (e){
handle catch
}


} )()

Reply:
Might have some further questions about this one, but it makes sense. I think I'm actually 'suffering' from this in #535, when the context updates and triggers the request, the rest of the useEffects run before the response, and then the updated state from the response doesn't trigger a render, so the state is updated but the rendered view is out of sync.